src/pstore/pstore.c

```diff
@@ -0,0 +1,407 @@
+                 * last six digits of the filename in determining the
+                 * record id.
+                 *
+                 * For the EFI backend, the record id encodes an id in the
+                 * upper 32 bits, and a timestamp in the lower 32-bits.
+                 * So ignoring the least significant 6 digits has proven
+                 * to generally identify related dmesg entries.  */
+#define PSTORE_FILENAME_IGNORE 6
+
+                /* determine common portion of record id */
+                ++p; /* move beyond dmesg- */
+                plen = strlen(p);
+                if (plen > PSTORE_FILENAME_IGNORE) {
+                        pe_id = malloc(plen - PSTORE_FILENAME_IGNORE + 1);
+                        if (!pe_id) {
+                                log_oom();
+                                return;
+                        }
+                        strncpy(pe_id, p, plen - PSTORE_FILENAME_IGNORE);
+                        pe_id[plen - PSTORE_FILENAME_IGNORE] = '\0';
+                } else
+                        pe_id = mfree(pe_id);
+
+                /* Now move file from pstore to archive storage */
+                move_file(pe, pe_id);
+
+                /* If the current record id is NOT the same as the
+                 * previous record id, then start a new dmesg.txt file */
+                if (!pe_id || !dmesg_id || !streq(pe_id, dmesg_id)) {
+                        /* Encountered a new dmesg group, close out old one, open new one */
+                        if (dmesg) {
+                                (void) write_dmesg(dmesg, dmesg_size, dmesg_id);
+                                dmesg = mfree(dmesg);
+                                dmesg_size = 0;
+                                dmesg_id = mfree(dmesg_id);
+                        }
+
+                        /* now point dmesg_id to storage of pe_id */
+                        dmesg_id = mfree(dmesg_id);
+                        dmesg_id = pe_id;
+                        pe_id = NULL; /* do not want this mfree'd below */
+                }
+
+                /* Reconstruction of dmesg is done as a useful courtesy, do not log errors */
+                dmesg = realloc(dmesg, dmesg_size + strlen(pe->dirent.d_name) + strlen(":\n") + pe->content_size + 1);
+                if (dmesg) {
+                        dmesg_size += sprintf(&dmesg[dmesg_size], "%s:\n", pe->dirent.d_name);
+                        if (pe->content) {
+                                memcpy(&dmesg[dmesg_size], pe->content, pe->content_size);
+                                dmesg_size += pe->content_size;
+                        }
+                }
+
+                pe_id = mfree(pe_id);
+        }
+        if (dmesg) {
+                (void) write_dmesg(dmesg, dmesg_size, dmesg_id);
+                dmesg = mfree(dmesg);
+                dmesg_size = 0;
+                dmesg_id = mfree(dmesg_id);
```

Comment:
`dmesg` and `dmesg_id` should be handled by `_cleanup_` anyway.

Reply:
done